README.md
# Crazy Turtle

This project is an attempt implement turtle graphics for the CrazyFlie 2.0
qaudcopter.  The goal is to enable users to control a Crazyflie's motion using
simple commands from a computer.

## Initial code base

The initial code is just experimental tests of the Crazyflie API.  Before we
can implement turtle graphics, we need need stable control of the quadcopter
from a compyuter.

## Licensing

All code for this project is licensed under the GNU General Public License
version 3 (GPLv3).  See the file COPYING for details.


NEWS
				 Crazy Turtle
				 ============

No news yet.


experiments/test/test.py
# -*- coding: utf-8 -*-
#
#     ||          ____  _ __
#  +------+      / __ )(_) /_______________ _____  ___
#  | 0xBC |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
#  +------+    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
#   ||  ||    /_____/_/\__/\___/_/   \__,_/ /___/\___/
#
#  Copyright (C) 2014 Bitcraze AB
#
#  Crazyflie Nano Quadcopter Client
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA  02110-1301, USA.

"""
Simple example that connects to the first Crazyflie found, ramps up/down
the motors and disconnects.
"""

import logging
import time, sys
from threading import Thread

#FIXME: Has to be launched from within the example folder
sys.path.append("../lib")
import cflib
from cfclient.utils.logconfigreader import LogConfig
from cflib.crazyflie import Crazyflie

import logging
logging.basicConfig(level=logging.ERROR)

class MotorRampExample:
    """Example that connects to a Crazyflie and ramps the motors up/down and
    the disconnects"""
    def __init__(self, link_uri):
        """ Initialize and run the example with the specified link_uri """

        self._cf = Crazyflie()

        self._cf.connected.add_callback(self._connected)
        self._cf.disconnected.add_callback(self._disconnected)
        self._cf.connection_failed.add_callback(self._connection_failed)
        self._cf.connection_lost.add_callback(self._connection_lost)

        self._cf.open_link(link_uri)

        print "Connecting to %s" % link_uri

    def _connected(self, link_uri):
        """ This callback is called form the Crazyflie API when a Crazyflie
        has been connected and the TOCs have been downloaded."""

        # Initially we go up and have done nothing with the barometer log
        self._thrust_mult = 1
        self._baro_count = 0;
        self._initial_baro_average = 0;
        self._baro_average = 0;

        # Start the barometer logging
        # The definition of the logconfig can be made before connecting
        self._lg_baro = LogConfig(name = "Baro", period_in_ms = 100);
        self._lg_baro.add_variable("baro.aslLong", "float");
        self._lg_baro.add_variable("stabilizer.roll", "float")
        self._lg_baro.add_variable("stabilizer.pitch", "float")
        self._lg_baro.add_variable("stabilizer.yaw", "float")
        self._lg_baro.add_variable("stabilizer.thrust", "uint16_t")

        # Initialize the stabilizer params
        self._roll = 0
        self._thrust = 0
        self._yaw = 0
        self._pitch = 0

        # Adding the configuration cannot be done until a Crazyflie is
        # connected, since we need to check that the variables we
        # would like to log are in the TOC.
        self._cf.log.add_config(self._lg_baro)
        if self._lg_baro.valid:
            # This callback will receive the data
            self._lg_baro.data_received_cb.add_callback(self._baro_log_data)
            # This callback will be called on errors
            self._lg_baro.error_cb.add_callback(self._baro_log_error)
            # Start the logging
            self._lg_baro.start()
        else:
            print("Could not add baro logconfig: some variables not in TOC")

        # Wait for the baseline barometer to be set
        time.sleep(1.0)

        # Start a separate thread to do the motor test.
        # Do not hijack the calling thread!
        Thread(target=self._ramp_motors).start()

    def _baro_log_error(self, logconf, msg):
        """Callback from the baro log API when an error occurs"""
        print "Error when logging baro %s: %s" % (logconf.name, msg)

    def _baro_log_data(self, timestamp, data, logconf):
        """Callback from the baro log API when data arrives"""
        # Initial barometer period is when we establish we are on the ground.
        # We calculate a rolling average of the barometer.
        self._baro_count = self._baro_count + 1

        print data
        self._roll = data['stabilizer.roll']
        self._thrust = data['stabilizer.thrust']
        self._yaw = data['stabilizer.yaw']
        self._pitch = data['stabilizer.pitch']

        if self._baro_count == 1:
            self._baro_average = data['baro.aslLong']
        else:
            self._baro_average = (self._baro_average * 0.9) + (data['baro.aslLong'] * 0.1)

        if self._baro_count <= 10:
            # Still in initial phase
            self._initial_baro_average = self._baro_average
        elif abs(self._initial_baro_average - self._baro_average) > 2.0:
            # In testing phase have climbed enough.
            self._thrust_mult = -1
            print "Flipping"

#        print "[%d][%s]: %f %f %f" % (timestamp, logconf.name,
#                                      data['baro.aslLong'],
#                                      self._initial_baro_average,
#                                      self._baro_average)

    def _connection_failed(self, link_uri, msg):
        """Callback when connection initial connection fails (i.e no Crazyflie
        at the speficied address)"""
        print "Connection to %s failed: %s" % (link_uri, msg)

    def _connection_lost(self, link_uri, msg):
        """Callback when disconnected after a connection has been made (i.e
        Crazyflie moves out of range)"""
        print "Connection to %s lost: %s" % (link_uri, msg)

    def _disconnected(self, link_uri):
        """Callback when the Crazyflie is disconnected (called in all cases)"""
        print "Disconnected from %s" % link_uri

    def _ramp_motors(self):
        thrust_step = 750
        thrust = 20000
        pitch = 0
        roll = 0
        yawrate = 0

        # New zeroes
        roll_zero  = -1.3407979011535645
        pitch_zero = -0.5004292726516724

        # Trim values
        roll_trim  = -roll_zero
        pitch_trim = -pitch_zero

        # offsets from the trim value
        roll_offset = 5.0
        pitch_offset = 5.0

        #Unlock startup thrust protection
        self._cf.commander.send_setpoint(0, 0, 0, 0)

        while thrust >= 20000:
        #while True:
            #if self._roll < roll_zero:
            #  roll = roll_trim + roll_offset
            #else:
            #  roll = roll_trim - roll_offset

            #if self._pitch < pitch_zero:
            #  pitch = pitch_trim + pitch_offset
            #else:
            #  pitch = pitch_trim - pitch_offset

            # calculate new values
            roll  = 0
            pitch = 0
            yaw = 170

            kill = False
            cutoff = 20.0
            cutoff_en = True

            if (cutoff_en and (abs(self._roll)  > cutoff or
                abs(self._pitch) > cutoff)):  #or
                #abs(self._yaw)   > cutoff)):
              kill = True

            if thrust >= 45000:
              kill = True

            # If kill is set, bring it down gracefully, else climb
            if not kill:
              print "Killswitch!"
              thrust += thrust_step
            else:
              thrust -= 2 * thrust_step

            print 'Current data', (self._roll, self._pitch,
                                  self._yaw,   self._thrust)
            print 'Sending data', (roll, pitch, yaw, thrust)

            self._cf.commander.send_setpoint(roll, pitch, yaw, thrust)
            time.sleep(0.1)

        self._cf.commander.send_setpoint(0, 0, 0, 0)
        # Make sure that the last packet leaves before the link is closed
        # since the message queue is not flushed before closing
        time.sleep(0.1)
        self._cf.close_link()

if __name__ == '__main__':
    # Initialize the low-level drivers (don't list the debug drivers)
    cflib.crtp.init_drivers(enable_debug_driver=False)
    # Scan for Crazyflies and use the first one found
    print "Scanning interfaces for Crazyflies..."
    available = cflib.crtp.scan_interfaces()
    print "Crazyflies found:"
    for i in available:
        print i[0]

    if len(available) > 0:
        le = MotorRampExample(available[0][0])
    else:
        print "No Crazyflies found, cannot run example"


experiments/jump/jump.py
# -*- coding: utf-8 -*-
#
#  Copyright (C) 2014 Bitcraze AB
#  Copyright (C) 2015 Embecosm Limited
#
#  Contributor  Jeremy Bennett <jeremy.bennett@embecosm.com>
#
#  Make the CrazyFlie hop.  Based on the ramp.py example from the Crazyflie
#  Nano Quadcopter Client
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA  02110-1301, USA.


"""
Module to make the Crazyflie hop

Simple example that connects to the first Crazyflie found, powers up the
motors for 500ms and disconnects.

Classes exported:
MotorJumpExample -- Instantiate to make the crazyflie hop
"""

import os
import logging
import time, sys
from threading import Thread

# If it's not on the path already, put the Crazyflie client library in the
# CFLIB environment variable.  This is the "lib" subdirectory of the Crazyflie
# client library.
try:
    sys.path.append(os.environ['CFLIB'])
except:
    # Just catch and ignore
    print "CFLIB not specified"

import cflib
from cfclient.utils.logconfigreader import LogConfig
from cflib.crazyflie import Crazyflie

import logging
logging.basicConfig(level=logging.ERROR)

class MotorJumpExample:
    """Class to make a Crazyflie hop

    There are no public classes or instance variables.  The behavior is
    invoked by instantiating the class.
    """

    def __init__(self, link_uri):
        """ Initialize and invoke behavior with the specified link_uri

        Keyword argument:
        link_uri -- the URI for the Crazyflie (no default)
        """

        # Instantiate a Crazyflie and set its callbacks
        self._cf = Crazyflie()

        self._cf.connected.add_callback(self._connected)
        self._cf.disconnected.add_callback(self._disconnected)
        self._cf.connection_failed.add_callback(self._connection_failed)
        self._cf.connection_lost.add_callback(self._connection_lost)

        # Log instantaneous data
        self._timestamp = 0
        self._asl       = 0.0
        self._roll      = 0.0
        self._pitch     = 0.0
        self._yaw       = 0.0
        self._thrust    = 0

        # Log rolling averages
        self._asl_av   = 0.0
        self._roll_av  = 0.0
        self._pitch_av = 0.0
        self._yaw_av   = 0.0

        # Rolling average is done as an infinite impulse response filter
        self._iir_coeff = 0.8

        # Opening the connection will cause the behavior to happen
        self._cf.open_link(link_uri)

        print "Connecting to %s" % link_uri


    def _connected(self, link_uri):
        """Callback on connection to a Crazyflie.

        This method is called from the Crazyflie API when a Crazyflie has been
        connected and the TOCs have been downloaded.

        It sets up some logging for the barometer, then after a short pause
        starts a separate thread to drive the motors.

        Keyword argument:
        link_uri -- the URI for the Crazyflie (no default)
        """

        # Start the logging.

        # The definition of the logconfig could have been made before
        # connecting, but we do it here, since the rest of the intialization
        # can't be done until we are connected.
        self._lg = LogConfig(name = "Log", period_in_ms = 20);
        self._lg.add_variable("baro.aslLong", "float");
        self._lg.add_variable("stabilizer.roll", "float")
        self._lg.add_variable("stabilizer.pitch", "float")
        self._lg.add_variable("stabilizer.yaw", "float")
        self._lg.add_variable("stabilizer.thrust", "uint16_t")

        self._first_datapoint = True;

        # Adding the configuration cannot be done until a Crazyflie is
        # connected, since we need to check that the variables we
        # would like to log are in the TOC.
        self._cf.log.add_config(self._lg)
        if self._lg.valid:
            # This callback will receive the data
            self._lg.data_received_cb.add_callback(self._log_data)
            # This callback will be called on errors
            self._lg.error_cb.add_callback(self._log_error)
            # Start the logging
            self._lg.start()
        else:
            print "Could not add logconfig"

        # Start a separate thread to do the motor test.
        # Do not hijack the calling thread!
        Thread(target=self._jump_motors).start()


    def _log_error(self, logconf, msg):
        """Callback from the log API when an error occurs

        We just note that an error has occurred.

        Keyword arguments:
        logconf -- configuration data for this log
        msg     -- message to print out.
        """

        print "Error when logging %s: %s" % (logconf.name, msg)


    def _log_data(self, timestamp, data, logconf):
        """Callback from the log API when data arrives

        We print out the data from the log.

        Keyword arguments:
        timestamp -- when the log data occurred
        data      -- the actual data
        logconf   -- configuration data for this log
        """

        #  Log data. Remember pitch is negative!
        self._asl = data['baro.aslLong']
        self._roll    = data['stabilizer.roll']
        self._pitch   = - data['stabilizer.pitch']
        self._yaw     = - data['stabilizer.yaw']
        self._thrust  = data['stabilizer.thrust']

        # Rolling averages. Start off with the value we have.
        self._timestamp   = timestamp
        if self._first_datapoint:
            self._asl_av  = self._asl
            self._roll_av     = self._roll
            self._pitch_av    = self._pitch
            self._yaw_av      = self._yaw

            self._first_datapoint = False;
        else:
            self._asl_av   = self._iir_filter(self._asl_av,   self._asl)
            self._roll_av  = self._iir_filter(self._roll_av,  self._roll)
            self._pitch_av = self._iir_filter(self._pitch_av, self._pitch)
            self._yaw_av   = self._iir_filter(self._yaw_av,   self._yaw)

        print "[%d]: %f (%f %f %f)"  % (timestamp,
                                        self._asl_av,
                                        self._roll_av,
                                        self._pitch_av,
                                        self._yaw_av)

    def _iir_filter(self, point, prev_res):
        """Apply the a simple infinite impulse response filter

          X[n] = (1-a).x[n] + a.X[n-1e]

        Keyword arguments:
        point    - the data point to add
        prev_res - the result prior to this point.

        Returns the updated IIR filter result
        """

        return (1.0 - self._iir_coeff) * point + prev_res * self._iir_coeff


    def _connection_failed(self, link_uri, msg):
        """Callback when connection initial connection fails

        We just print out a message.

        Keyword arguments:
        link_uri -- the URI for the Crazyflie
        msg      -- message to print out
        """
        print "Connection to %s failed: %s" % (link_uri, msg)


    def _connection_lost(self, link_uri, msg):
        """Callback when disconnected after a connection has been made

        This typically occurs if Crazyflie moves out of range.  We just print
        a message.

        Keyword arguments:
        link_uri -- the URI for the Crazyflie
        msg      -- message to print out
        """
        print "Connection to %s lost: %s" % (link_uri, msg)


    def _disconnected(self, link_uri):
        """Callback when the Crazyflie is disconnected

        This is normal behaviour, and we just print a message.

        Keyword arguments:
        link_uri -- the URI for the Crazyflie
        """
        print "Disconnected from %s" % link_uri


    def _jump_motors(self):
        """Make the Crazyflie jump

        Just a burst of the motors.
        """

        # Kill parameters
        kill = False
        cutoff = 20.0
        cutoff_en = True

        # Parameters
        # Roll - negative makes it go left
        # Pitch - positive makes it go forward
        thrust_step  =   500
        thrust_start = 20000
        thrust_stop  = 37000
        thrust_limit = 50000

        # Target as a fraction of the take off ASL
        asl_frac = 0.995

        roll  = -2
        pitch = 5
        yaw   = 0

        step_period = 0.05
        jumptime = 3
        jumpsteps = int(float(jumptime) / step_period)

        # Unlock startup thrust protection
        self._cf.commander.send_setpoint(0, 0, 0, 0)

        # Rapidly wind up
        for t in range(thrust_start, thrust_stop, thrust_step):
            self._cf.commander.send_setpoint(roll, pitch, yaw, t)
            time.sleep(0.01)

        self._aslTarget = self._asl * asl_frac
        print "Target ASL %f" % (self._aslTarget)

        thrust = thrust_stop
        dircount = 0
        prev_asl = self._asl_av

        # Then new figures every second
        for i in range (1, jumpsteps, 1):
            if (cutoff_en and (abs(self._roll_av)  > cutoff or
                               abs(self._pitch_av) > cutoff)):
                print "Cutoff (%f %f)" % (self._roll_av, self._pitch_av)
                kill = True

            if kill:
                # Die politely
                print "Kill switch"
                self._cf.commander.send_setpoint(0, 0, 0, 0)
                time.sleep(1.0)
                self._cf.close_link()
            else:
                print "Correcting: (%f %f %f) %d"  % (roll  - self._roll_av,
                                                      pitch - self._pitch_av,
                                                      yaw   - self._yaw_av,
                                                      thrust)

                self._cf.commander.send_setpoint(roll  - self._roll_av,
                                                 pitch - self._pitch_av,
                                                 yaw   - self._yaw_av,
                                                 thrust)

            # Wait a little bit
            time.sleep(step_period)

        # Turn the motors off
        self._cf.commander.send_setpoint(0, 0, 0, 0)

        # Make sure that the last packet leaves before the link is closed
        # since the message queue is not flushed before closing
        time.sleep(1.0)
        self._cf.close_link()


# Initialize the instance to make the motor jump
if __name__ == '__main__':
    # Initialize the low-level drivers (don't list the debug drivers)
    cflib.crtp.init_drivers(enable_debug_driver=False)
    # Scan for Crazyflies and use the first one found
    print "Scanning interfaces for Crazyflies..."
    available = cflib.crtp.scan_interfaces()
    print "Crazyflies found:"
    for i in available:
        print i[0]

    if len(available) > 0:
        le = MotorJumpExample(link_uri = available[0][0])
    else:
        print "No Crazyflies found, cannot run example"


experiments/basiclog/basiclog.py
# -*- coding: utf-8 -*-
#
#     ||          ____  _ __
#  +------+      / __ )(_) /_______________ _____  ___
#  | 0xBC |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
#  +------+    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
#   ||  ||    /_____/_/\__/\___/_/   \__,_/ /___/\___/
#
#  Copyright (C) 2014 Bitcraze AB
#
#  Crazyflie Nano Quadcopter Client
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA  02110-1301, USA.

"""
Simple example that connects to the first Crazyflie found, logs the Stabilizer
and prints it to the console. After 10s the application disconnects and exits.
"""

import sys
sys.path.append("../lib")

import cflib.crtp

import logging
import time
from threading import Timer

import cflib.crtp
from cfclient.utils.logconfigreader import LogConfig
from cflib.crazyflie import Crazyflie

# Only output errors from the logging framework
logging.basicConfig(level=logging.ERROR)

class LoggingExample:
    """
    Simple logging example class that logs the Stabilizer from a supplied
    link uri and disconnects after 5s.
    """
    def __init__(self, link_uri):
        """ Initialize and run the example with the specified link_uri """

        # Create a Crazyflie object without specifying any cache dirs
        self._cf = Crazyflie()

        # Connect some callbacks from the Crazyflie API
        self._cf.connected.add_callback(self._connected)
        self._cf.disconnected.add_callback(self._disconnected)
        self._cf.connection_failed.add_callback(self._connection_failed)
        self._cf.connection_lost.add_callback(self._connection_lost)

        print "Connecting to %s" % link_uri

        # Try to connect to the Crazyflie
        self._cf.open_link(link_uri)

        # Variable used to keep main loop occupied until disconnect
        self.is_connected = True

    def _connected(self, link_uri):
        """ This callback is called form the Crazyflie API when a Crazyflie
        has been connected and the TOCs have been downloaded."""
        print "Connected to %s" % link_uri

        # The definition of the logconfig can be made before connecting
#        self._lg_stab = LogConfig(name="Stabilizer", period_in_ms=100)
#        self._lg_stab.add_variable("stabilizer.roll", "float")
#        self._lg_stab.add_variable("stabilizer.pitch", "float")
#        self._lg_stab.add_variable("stabilizer.yaw", "float")
#        self._lg_stab.add_variable("stabilizer.thrust", "uint16_t")

        self._lg_baro = LogConfig(name = "Baro", period_in_ms = 100);
        self._lg_baro.add_variable("baro.aslLong", "float");

        # Adding the configuration cannot be done until a Crazyflie is
        # connected, since we need to check that the variables we
        # would like to log are in the TOC.
#        self._cf.log.add_config(self._lg_stab)
#        if self._lg_stab.valid:
#            # This callback will receive the data
#            self._lg_stab.data_received_cb.add_callback(self._stab_log_data)
#            # This callback will be called on errors
#            self._lg_stab.error_cb.add_callback(self._stab_log_error)
#            # Start the logging
#            self._lg_stab.start()
#        else:
#            print("Could not add stablizer logconfig since some variables are not in TOC")

        self._cf.log.add_config(self._lg_baro)
        if self._lg_baro.valid:
            # This callback will receive the data
            self._lg_baro.data_received_cb.add_callback(self._baro_log_data)
            # This callback will be called on errors
            self._lg_baro.error_cb.add_callback(self._baro_log_error)
            # Start the logging
            self._lg_baro.start()
        else:
            print("Could not add baro logconfig since some variables are not in TOC")

        # Start a timer to disconnect in 10s
        t = Timer(5, self._cf.close_link)
        t.start()

    def _stab_log_error(self, logconf, msg):
        """Callback from the stabilizer log API when an error occurs"""
        print "Error when logging stabilizer %s: %s" % (logconf.name, msg)

    def _stab_log_data(self, timestamp, data, logconf):
        """Callback from the stablizer log API when data arrives"""
        print "[%d][%s]: %s" % (timestamp, logconf.name, data)

    def _baro_log_error(self, logconf, msg):
        """Callback from the baro log API when an error occurs"""
        print "Error when logging baro %s: %s" % (logconf.name, msg)

    def _baro_log_data(self, timestamp, data, logconf):
        """Callback froma the baro log API when data arrives"""
        print "[%d][%s]: %s" % (timestamp, logconf.name, data)

    def _connection_failed(self, link_uri, msg):
        """Callback when connection initial connection fails (i.e no Crazyflie
        at the speficied address)"""
        print "Connection to %s failed: %s" % (link_uri, msg)
        self.is_connected = False

    def _connection_lost(self, link_uri, msg):
        """Callback when disconnected after a connection has been made (i.e
        Crazyflie moves out of range)"""
        print "Connection to %s lost: %s" % (link_uri, msg)

    def _disconnected(self, link_uri):
        """Callback when the Crazyflie is disconnected (called in all cases)"""
        print "Disconnected from %s" % link_uri
        self.is_connected = False

if __name__ == '__main__':
    # Initialize the low-level drivers (don't list the debug drivers)
    cflib.crtp.init_drivers(enable_debug_driver=False)
    # Scan for Crazyflies and use the first one found
    print "Scanning interfaces for Crazyflies..."
    available = cflib.crtp.scan_interfaces()
    print "Crazyflies found:"
    for i in available:
        print i[0]

    if len(available) > 0:
        le = LoggingExample(available[0][0])
    else:
        print "No Crazyflies found, cannot run example"

    # The Crazyflie lib doesn't contain anything to keep the application alive,
    # so this is where your application should do something. In our case we
    # are just waiting until we are disconnected.
    while le.is_connected:
        time.sleep(1)


experiments/ramp/ramp.py
# -*- coding: utf-8 -*-
#
#     ||          ____  _ __
#  +------+      / __ )(_) /_______________ _____  ___
#  | 0xBC |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
#  +------+    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
#   ||  ||    /_____/_/\__/\___/_/   \__,_/ /___/\___/
#
#  Copyright (C) 2014 Bitcraze AB
#
#  Crazyflie Nano Quadcopter Client
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; either version 2
#  of the License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.

#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA  02110-1301, USA.

"""
Simple example that connects to the first Crazyflie found, ramps up/down
the motors and disconnects.
"""

import logging
import time, sys
from threading import Thread

#FIXME: Has to be launched from within the example folder
sys.path.append("../lib")
import cflib
from cfclient.utils.logconfigreader import LogConfig
from cflib.crazyflie import Crazyflie

import logging
logging.basicConfig(level=logging.ERROR)

class MotorRampExample:
    """Example that connects to a Crazyflie and ramps the motors up/down and
    the disconnects"""
    def __init__(self, link_uri):
        """ Initialize and run the example with the specified link_uri """

        self._cf = Crazyflie()

        self._cf.connected.add_callback(self._connected)
        self._cf.disconnected.add_callback(self._disconnected)
        self._cf.connection_failed.add_callback(self._connection_failed)
        self._cf.connection_lost.add_callback(self._connection_lost)

        self._cf.open_link(link_uri)

        print "Connecting to %s" % link_uri

    def _connected(self, link_uri):
        """ This callback is called form the Crazyflie API when a Crazyflie
        has been connected and the TOCs have been downloaded."""

        # Initially we go up and have done nothing with the barometer log
        self._thrust_mult = 1
        self._baro_count = 0;
        self._initial_baro_average = 0;
        self._baro_average = 0;

        # Start the barometer logging
        # The definition of the logconfig can be made before connecting
        self._lg_baro = LogConfig(name = "Baro", period_in_ms = 100);
        self._lg_baro.add_variable("baro.aslLong", "float");

        # Adding the configuration cannot be done until a Crazyflie is
        # connected, since we need to check that the variables we
        # would like to log are in the TOC.
        self._cf.log.add_config(self._lg_baro)
        if self._lg_baro.valid:
            # This callback will receive the data
            self._lg_baro.data_received_cb.add_callback(self._baro_log_data)
            # This callback will be called on errors
            self._lg_baro.error_cb.add_callback(self._baro_log_error)
            # Start the logging
            self._lg_baro.start()
        else:
            print("Could not add baro logconfig: some variables not in TOC")

        # Wait for the baseline barometer to be set
        time.sleep(1.0)

        # Start a separate thread to do the motor test.
        # Do not hijack the calling thread!
        Thread(target=self._ramp_motors).start()

    def _baro_log_error(self, logconf, msg):
        """Callback from the baro log API when an error occurs"""
        print "Error when logging baro %s: %s" % (logconf.name, msg)

    def _baro_log_data(self, timestamp, data, logconf):
        """Callback from the baro log API when data arrives"""
        # Initial barometer period is when we establish we are on the ground.
        # We calculate a rolling average of the barometer.
        self._baro_count = self._baro_count + 1

        if self._baro_count == 1:
            self._baro_average = data['baro.aslLong']
        else:
            self._baro_average = (self._baro_average * 0.9) + (data['baro.aslLong'] * 0.1)

        if self._baro_count <= 10:
            # Still in initial phase
            self._initial_baro_average = self._baro_average
        elif abs(self._initial_baro_average - self._baro_average) > 2.0:
            # In testing phase have climbed enough.
            self._thrust_mult = -1
            print "Flipping"

#        print "[%d][%s]: %f %f %f" % (timestamp, logconf.name,
#                                      data['baro.aslLong'],
#                                      self._initial_baro_average,
#                                      self._baro_average)

    def _connection_failed(self, link_uri, msg):
        """Callback when connection initial connection fails (i.e no Crazyflie
        at the speficied address)"""
        print "Connection to %s failed: %s" % (link_uri, msg)

    def _connection_lost(self, link_uri, msg):
        """Callback when disconnected after a connection has been made (i.e
        Crazyflie moves out of range)"""
        print "Connection to %s lost: %s" % (link_uri, msg)

    def _disconnected(self, link_uri):
        """Callback when the Crazyflie is disconnected (called in all cases)"""
        print "Disconnected from %s" % link_uri

    def _ramp_motors(self):
        thrust_step = 500
        thrust = 20000
        pitch = 0
        roll = 0
        yawrate = 0

        #Unlock startup thrust protection
        self._cf.commander.send_setpoint(0, 0, 0, 0)

        while thrust >= 20000:
            self._cf.commander.send_setpoint(roll, pitch, yawrate, thrust)
            time.sleep(0.1)
            if thrust >= 40000:
                self._thrust_mult = -1
                print "Flipped safety"
            thrust += thrust_step * self._thrust_mult
            print "thrust %d" % (thrust)
        self._cf.commander.send_setpoint(0, 0, 0, 0)
        # Make sure that the last packet leaves before the link is closed
        # since the message queue is not flushed before closing
        time.sleep(0.1)
        self._cf.close_link()

if __name__ == '__main__':
    # Initialize the low-level drivers (don't list the debug drivers)
    cflib.crtp.init_drivers(enable_debug_driver=False)
    # Scan for Crazyflies and use the first one found
    print "Scanning interfaces for Crazyflies..."
    available = cflib.crtp.scan_interfaces()
    print "Crazyflies found:"
    for i in available:
        print i[0]

    if len(available) > 0:
        le = MotorRampExample(available[0][0])
    else:
        print "No Crazyflies found, cannot run example"
